---
title: "assignment3"
author: "Francesco Bianchi"
date: "2023-03-15"
output: html_document
---

```{r}
suppressPackageStartupMessages({
  library(AnnotationHub)
  library(Rsubread)
  library(rtracklayer)
  library(Biostrings)
  library(Rfastp)
  library(epiwraps)
})
ah <- AnnotationHub()
```

```{r}
getwd()
```

*1. Download the following Drosophila ChIP-seq for the protein CTCF:

```{r, eval=FALSE}
#increase download times for large downloads
options(timeout=3600)
#creation of subdirectory
dir.create("raw")
download.file("https://www.encodeproject.org/files/ENCFF127RRR/@@download/ENCFF127RRR.fastq.gz", dest="raw/Myc.fastq.gz")
```
*2. Process it from the raw data, obtaining:
**a. bam file

#Aligning reads to create .bam files

## Using Rsubread

### Building a genome index for mapping

This we have to do only once for a genome, and can then re-use across projects
the trimfastq files look exactly like the normal fastq files, but are smaller

```{r, eval=FALSE}
# we get the genome sequence from AnnotationHub
genome <- ah[["AH49674"]]
# we create a new directory that will contain the genome index
dir.create("BDGP6_genome")
# we write the genome sequence in fasta format
export(import.2bit(genome), "BDGP6_genome/genome.fasta.gz", compress=TRUE)
# we build a Rsubread index (can also be indexed by other software of choice)
Rsubread::buildindex("BDGP6_genome/rsubread", reference="BDGP6_genome/genome.fasta.gz")
```

### Alignment
```{r}
dir.create("aligned_A")
align.stats <- Rsubread::align(index="BDGP6_genome/rsubread", type="dna",
                               readfile1=c("rfastp.trimmed/Myc_R1.fastq.gz", 
                                           "rfastp.trimmed/input_R1.fastq.gz"),
                               output_file=c("aligned/Myc.bam","aligned/input.bam"),
                               nthreads=6, sortReadsByCoordinates=TRUE)
```

### Trimming
especially important in epigenomics data, not so much in RNA data
## Using Rfastp
(no input control needed for this exercise)

```{r}
dir.create("trimmed")

fastq_files <- c(Myc="raw/Myc.fastq.gz", input="raw/input.fastq.gz")


# package::function specifies that a certain function is from a certain package (for functions with identical names)
align.stats <- Rsubread::align(index="BDGP6_genome/rsubread", type="dna",
                               readfile1=c("rfastp.trimmed/Myc_R1.fastq.gz"),
                               output_file=c("aligned/Myc.bam"),
                               nthreads=6, sortReadsByCoordinates=TRUE)
#alignment statistics e.g., total reads, unmapped reads etc.
align.stats
```


**b. peaks
```{r}
peaks <- callPeaks("aligned/Myc.bam", fragLength=50L)
# if we want to save it as a bed file:
#rtracklayer::export.bed(peaks, "peaks/peaks.bed")
```



*3. Report:
**a. how many reads (and what percentage) were mapped
7446729	 out of 16109551 reads were mapped, which is equivalent to 46%
**b. how many peaks were found
Number of peaks found is 4833 peaks after filtering
*4. Plot the signal around one of the peaks

```{r, eval=FALSE}
peaks[1]
plotSignalTracks(files = c(Myc="aligned/Myc.bam", Input="aligned/input.bam"), region=peaks[1], extend=1000)
```