---
title: "Assignment 07"
author: "Frncesco Bianchi"
date: "2024-04-25"
output: html_document
---

```{r Packages}
suppressPackageStartupMessages({
  library(epiwraps)
  library(AnnotationHub)
  library(MotifDb)
# library(memes)
  library(universalmotif)
  library(ensembldb)
  library(ggplot2)
# library(magick)
})
# update epiwraps
BiocManager::install("ETHZ-INS/epiwraps")

ah <- AnnotationHub(localHub=FALSE)
ensdb <- ah[["AH89211"]]

```



## Download the data


```{r, eval=FALSE}
BiocManager::install("ETHZ-INS/epiwraps")
download.file("https://ethz-ins.org/content/w7/atac.chr19.bam", "atac.chr19.bam")
download.file("https://ethz-ins.org/content/w7/atac.chr19.bam.bai", "atac.chr19.bam.bai")
```

## Preparing tracks


```{r create Tracks}
bam <- "atac.chr19.bam"

# 1. Create a track using only nucleosome-free fragments, the number of cuts/insertion sites at each position
epiwraps::bam2bw(bam, output_bw = "NF_cuts.bw", paired=TRUE, binWidth=1L, type="ends", extend=2L, minFragLength=30, 
       maxFragLength=120, shift=c(4L,-5L), forceSeqlevelsStyle = "Ensembl")
# Load the insertion profile from the BigWig file
track_NF_cuts <- rtracklayer::import("NF_cuts.bw")

# 2. Create a track using only the (10bp) centers of mono-nucleosome fragments. Center and extend is to get rid of junk.
epiwraps::bam2bw(bam, output_bw = "mono_centers.bw", paired=TRUE, binWidth=5L, minFragLength=140, shift=c(4L,-5L), 
       maxFragLength=220, type="center", extend=10L, forceSeqlevelsStyle = "Ensembl")
# Load the insertion profile from the BigWig file
track_mono_centers <- rtracklayer::import("mono_centers.bw")
```

## Obtaining the sites with motifs (KLF4 and MAZ)


```{r}
# get KLF4 motif
motif <- MotifDb::query(MotifDb, c("KLF4","Mus"))[[1]]
motif_KLF4 <- convert_motifs(motif, class="TFBSTools-PFMatrix")

# get MAZ motif
motif <- MotifDb::query(MotifDb, c("MAZ","Mus"))[[1]]
motif_MAZ <- convert_motifs(motif, class="TFBSTools-PFMatrix")

genome <- ah[["AH68356"]]
# get the sequence for chr19:
chr19 <- import(genome)["19"]

# find motif matches of KLF4 across chr19
moi_KLF4 <- motifmatchr::matchMotifs(motif_KLF4, chr19, out="positions", p.cutoff=1e-5)[[1]]
# convert to GRanges
moi_KLF4 <- as(setNames(moi_KLF4,names(chr19)), "GRanges")

# find motif matches of MAZ across chr19
moi_MAZ <- motifmatchr::matchMotifs(motif_MAZ, chr19, out="positions", p.cutoff=1e-5)[[1]]
# convert to GRanges
moi_MAZ <- as(setNames(moi_MAZ,names(chr19)), "GRanges")
```

# Plots of signals around the motifs

### KLF4


```{r KLF4}
# we prepare the list of tracks
tracks_KLF4 <- c("NF cuts"="NF_cuts.bw", "Mono centers"="mono_centers.bw")

# extract signals around the motif occurences
# we zoom in to 300bp around the motif centers, in windows of 5bp
sm_KLF4 <- signal2Matrix(tracks_KLF4, moi_KLF4, w=5, extend=300)

# plot the signals:

# background normalization
# this we can do for instance using:
nf_KLF4 <- getNormFactors(tracks_KLF4, useSeqLevels="19", nwind=5000L)
# then we apply the normalization factors:
sm_KLF4 <- renormalizeSignalMatrices(sm_KLF4, scaleFactors = nf_KLF4)
# and plot it
plotEnrichedHeatmaps(sm_KLF4, trim=0.95, minRowVal = 12, colors = c("white","darkred"), use_raster = FALSE)
```

### MAZ


```{r MAZ}
# we prepare the list of tracks
tracks_MAZ <- c("NF cuts"="NF_cuts.bw", "Mono centers"="mono_centers.bw")

# extract signals around the motif occurences
# we zoom in to 300bp around the motif centers, in windows of 5bp
sm_MAZ <- signal2Matrix(tracks_MAZ, moi_MAZ, w=5, extend=300)

# plot the signals:
# plotEnrichedHeatmaps(sm_MAZ, trim=0.95, use_raster = FALSE) # not the proper way

# background normalization
nf_MAZ <- getNormFactors(tracks_MAZ, useSeqLevels="19", nwind=5000L)
# then we apply the normalization factors:
sm_MAZ <- renormalizeSignalMatrices(sm_MAZ, scaleFactors = nf_MAZ)
# and plot it
plotEnrichedHeatmaps(sm_MAZ, trim=0.95, minRowVal = 12, colors = c("white","darkred"), use_raster = FALSE)
```

#Interpretation
#KLF4:
##For 'NF cuts', the heat map and density plot appear to show a moderate signal without strong peaks. This indicates a relatively uniform or low enrichment in the regions. For ‘mono-centres’, the heatmap shows a stronger enhancement (darker colour) around the centre and the density plot confirms a peak at this point. This implies that what is being measured is concentrated in the centres of these regions.

##MAZ
#'NF cuts' show a distributed or less specific pattern of the genomic feature in the regions, indicating that the signal or feature is not specifically located in the centre, but is possibly distributed over a larger area or is only present to a lesser extent overall.'Mono centres' show a very strong central enrichment, indicating that the analysed trait is very specifically located at these central points in the genome or occurs there. This could indicate strict regulation or strong specificity of protein binding, DNA accessibility or the presence of specific markers in these regions.

